File: backend/app/schemas/pnl.py
"""P&L report schemas."""
from decimal import Decimal
from typing import Optional
import uuid

from pydantic import BaseModel, ConfigDict


class PnLLineItems(BaseModel):
    """All P&L line items. None when data is unavailable for a period."""

    gross_revenue: Optional[Decimal] = None
    total_discounts: Optional[Decimal] = None
    net_revenue: Optional[Decimal] = None

    cogs: Optional[Decimal] = None
    gross_profit: Optional[Decimal] = None

    labor_cost: Optional[Decimal] = None
    prime_cost: Optional[Decimal] = None

    operating_expenses: Optional[Decimal] = None

    ebitda: Optional[Decimal] = None
    net_profit: Optional[Decimal] = None

    # Percentage breakdowns (of net_revenue)
    cogs_pct: Optional[Decimal] = None
    labor_pct: Optional[Decimal] = None
    prime_cost_pct: Optional[Decimal] = None
    ebitda_pct: Optional[Decimal] = None
    net_profit_pct: Optional[Decimal] = None


class ExpenseLineItem(BaseModel):
    vendor_name: Optional[str] = None
    amount: Decimal


class ExpenseCategoryBreakdown(BaseModel):
    category: str
    total: Decimal
    expense_count: int
    expenses: list[ExpenseLineItem] = []


class PnLReportResponse(BaseModel):
    """Full P&L report for a date range."""

    tenant_id: uuid.UUID
    location_id: Optional[uuid.UUID] = None
    period_start: str
    period_end: str
    currency_code: str = "CAD"

    line_items: PnLLineItems
    expense_breakdown: list[ExpenseCategoryBreakdown] = []

    order_count: int = 0
    expense_count: int = 0

    # True only when at least one bank_statement document's date falls inside
    # the period. P&L numbers are computed either way — this just flags that
    # nothing has verified the expense/revenue data against the bank record.
    bank_statement_verified: bool = False
    bank_statement_warning: Optional[str] = None


class DailyRevenuePoint(BaseModel):
    date: str  # YYYY-MM-DD
    gross_revenue: Decimal
    net_revenue: Decimal
    void_amount: Decimal
    order_count: int


class DailyBreakdownResponse(BaseModel):
    period_start: str
    period_end: str
    points: list[DailyRevenuePoint] = []


class PnLSnapshotResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    tenant_id: uuid.UUID
    location_id: Optional[uuid.UUID] = None
    period_start: str
    period_end: str
    period_label: str
    gross_revenue: Optional[Decimal] = None
    total_discounts: Optional[Decimal] = None
    net_revenue: Optional[Decimal] = None
    cogs: Optional[Decimal] = None
    gross_profit: Optional[Decimal] = None
    labor_cost: Optional[Decimal] = None
    prime_cost: Optional[Decimal] = None
    operating_expenses: Optional[Decimal] = None
    ebitda: Optional[Decimal] = None
    net_profit: Optional[Decimal] = None
    order_count: int
    expense_count: int

File: frontend/src/types/pnl.ts
export interface PnLLineItems {
  gross_revenue: string | null;
  total_discounts: string | null;
  net_revenue: string | null;
  cogs: string | null;
  gross_profit: string | null;
  labor_cost: string | null;
  prime_cost: string | null;
  operating_expenses: string | null;
  ebitda: string | null;
  interest_expense: string | null;
  net_profit: string | null;
  cogs_pct: string | null;
  labor_pct: string | null;
  prime_cost_pct: string | null;
  ebitda_pct: string | null;
  net_profit_pct: string | null;
}

export type BenchmarkStatus = "good" | "watch" | "bad" | "unknown";

export interface BenchmarkChip {
  metric: string;
  label: string;
  value_pct: string | null;
  target_label: string;
  status: BenchmarkStatus;
}

export interface DiscountBreakdownRow {
  name: string;
  scope: string;
  total: string;
  count: number;
  pct_of_discounts: string | null;
}

export interface DiscountBreakdown {
  total: string;
  currency_code: string;
  discounts: DiscountBreakdownRow[];
}

export interface PnLTrendPoint {
  period_label: string;
  period_start: string;
  net_revenue: string | null;
  cogs: string | null;
  cogs_pct: string | null;
  labor_cost: string | null;
  labor_pct: string | null;
  prime_cost_pct: string | null;
  ebitda: string | null;
  net_profit: string | null;
  net_profit_pct: string | null;
}

export interface PnLTrend {
  months: number;
  points: PnLTrendPoint[];
}

export interface ExpenseLineItem {
  vendor_name: string | null;
  amount: string;
}

export interface ExpenseCategoryBreakdown {
  category: string;
  total: string;
  expense_count: number;
  expenses: ExpenseLineItem[];
}

export interface PnLReport {
  tenant_id: string;
  location_id: string | null;
  period_start: string;
  period_end: string;
  currency_code: string;
  line_items: PnLLineItems;
  expense_breakdown: ExpenseCategoryBreakdown[];
  benchmarks: BenchmarkChip[];
  order_count: number;
  expense_count: number;
  bank_statement_verified: boolean;
  bank_statement_warning: string | null;
}

export interface PnLSnapshot {
  id: string;
  tenant_id: string;
  location_id: string | null;
  period_start: string;
  period_end: string;
  period_label: string;
  gross_revenue: string | null;
  net_revenue: string | null;
  cogs: string | null;
  gross_profit: string | null;
  labor_cost: string | null;
  ebitda: string | null;
  net_profit: string | null;
  order_count: number;
  expense_count: number;
}

// ─── Bank-statement-basis P&L ────────────────────────────────────────────────

export interface BankBasisSide {
  revenue: string;
  ebitda: string;
  net_profit: string;
}

export interface BankPartnerSplit {
  name: string;
  share_pct: string;
  revenue_before_hst: string;
  revenue_after_hst: string;
  net_before_hst: string;
  net_after_hst: string;
}

export interface BankBasisPnL {
  period_start: string;
  period_end: string;
  basis: string;
  revenue: string;
  revenue_by_channel: Record<string, string>;
  cogs: string;
  labor: string;
  operating_expenses: string;
  interest: string;
  expense_by_category: Record<string, string>;
  principal_excluded: string;
  before_hst: BankBasisSide;
  after_hst: BankBasisSide;
  hst: {
    collected_on_sales: string;
    input_tax_credits: string;
    net_remittance: string;
  };
  partner_split: BankPartnerSplit[];
}

export interface Partner {
  name: string;
  share_pct: string;
}
